wandb/sdk/launch/runner/kubernetes.py

```diff
@@ -0,0 +1,222 @@
+
+        # extract job spec component parts for convenience
+        job_metadata = job_dict.get("metadata", {})
+        job_spec = job_dict.get("spec", {})
+        pod_template = job_spec.get("template", {})
+        pod_metadata = pod_template.get("metadata", {})
+        pod_spec = pod_template.get("spec", {})
+        containers = pod_spec.get("containers", [{}])
+        job_status = job_dict.get('status', {})
+
+        # begin pulling resource arg overrides. all of these are optional
+
+        # allow top-level namespace override, otherwise take namespace specified at the job level, or default
+        namespace = resource_args.get("namespace", job_metadata.get("namespace", "default"))
+
+        # name precedence: resource args override > name in spec file > generated name
+        job_metadata['name'] = resource_args.get("job_name", job_metadata.get('name'))
+        if not job_metadata.get('name'):
+            job_metadata['generateName'] = "launch-"
+
+        if resource_args.get('cluster_name'):
+            job_metadata['clusterName'] = resource_args.get("cluster_name")
+        if resource_args.get('job_labels'):
+            job_metadata['labels'] = resource_args.get('job_labels')
+
+        if resource_args.get('backoff_limit'):
+            job_spec['backoffLimit'] = resource_args.get('backoff_limit')
+        if resource_args.get('job_completions'):
+            job_spec['completions'] = resource_args.get('job_completions')
+        if resource_args.get('parallelism'):
+            job_spec['parallelism'] = resource_args.get('parallelism')
+        if resource_args.get('suspend'):
+            job_spec['suspend'] = resource_args.get('suspend')
+
+        # only support container overrides for the single container case
+        if any(arg in resource_args for arg in ['container_name', 'resource_requests', 'resource_limits']):
+            raise LaunchError("Resource overrides not supported for multiple containers. Multiple container configurations should be specified in a yaml file supplied via job_spec.")
+
+        containers[0]['name'] = resource_args.get("container_name", containers[0].get('name', 'launch'))
+        container_resources = containers[0].get('resources', {})
+        if resource_args.get('resource_requests'):
+            container_resources['requests'] = resource_args.get("resource_requests")
+        if resource_args.get('resource_limits'):
+            container_resources['limits'] = resource_args.get('resource_limits')
+        if container_resources:
+            containers[0]['resources'] = container_resources
+        # todo: args and env vars would be added here, need to figure out what kind of overrides we want
+
+        pod_spec['restartPolicy'] = resource_args.get('pod_restart_policy', 'Never')
+        if resource_args.get('pod_preemption_policy'):
+            pod_spec['preemptionPolicy'] = resource_args.get("pod_preemption_policy")
```

Comment:
are premptionPolicuy, nodeName and nodeSelectors optional?

Reply:
Yep